enhancements/cloud-integration/credential-rotation.md

```diff
@@ -0,0 +1,183 @@
+approvers:
+  - "@sdodson"
+  - "@derekwaynecarr"
+creation-date: 2020-10-06
+last-updated: 2020-10-08
+status: provisional
+---
+
+# Cloud Credential Rotation
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Implement automatic credentials rotation in the cloud-credential-operator.
+Allow users to configure a global expiry duration for their cloud credentials config,
+after which the CCO will rotate new credentials in, and cleanup the old after a
+suitable period of time.
+
+## Motivation
+
+Customers often have security policies that prohibit the use of long lived
+cloud credentials. This enhancement provides a solution to a frequently
+requested feature, will improve OpenShift security overall, and help meet the
+requirements of more customers and thus generate more sales.
+
+### Goals
+
+  1. Support a global expiry time for CredentialsRequest credentials, defined in the CloudCredential config CR.
+  1. Rotate expired credentials when expiry time has been reached.
+  1. Preserve previous credentials for a period of time to ensure operators remain fully functional.
+  1. All operators using CredentialsRequests able to successfully handle a Secret modification with new credentials without going degraded.
+
+### Non-Goals
+
+  1. Mint mode only, not applicable to Passthrough or Manual mode. (although the operator support for credentials changing will be universally useful in both of these modes for customers manually rotating their own credentials)
+  1. AWS STS Token support. This is separate from this effort and not applicable to all mint capable cloud providers.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I would like to be able to define a credentials
+expiration policy and have my OpenShift CredentailsRequests rotated
+automatically.
+
+#### Story 2
+
+As a cluster administrator, I would like OpenShift operators to gracefully
+detect new credentials and not go degraded during the transition.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This implementation is only relevant for cloud actuators that support mint
+mode. (presently AWS, Azure, and GCP) Expiration makes no sense for Passthrough
+or Manual mode, where users can control credentials rotation themselves.
+
+#### API Changes
+
+Add CloudCredential.Spec.ExpireHours (int64) to represent the global setting
+for how long credentials should live before being rotated. If this is specified
+when the operator is *not* in mint mode, this should be reported as a Degraded
+state. It is assumed that hours is the most reasonable granularity for
+rotation.
+
+Add CredentialsRequest.Status.MintTime timestamp to store the last time
```

Comment:
Getting into implementation, while the kube api guidelines say to use Time not Timestamp, we're already using Timestamp in the Status on this resource and it's probably best to stay consistent.

Would RotationTimestamp be more or less accurate? I wouldn't want to indicate that this would be set for Passthrough mode.